asserts/model.go

```diff
@@ -25,16 +25,310 @@ import (
 	"strings"
 	"time"
 
+	"github.com/snapcore/snapd/snap/channel"
 	"github.com/snapcore/snapd/snap/naming"
 	"github.com/snapcore/snapd/strutil"
 )
 
+// ModelSnap holds the details about a snap specified by a model assertion.
+type ModelSnap struct {
+	Name   string
+	SnapID string
+	// SnapType is one of: app|base|gadget|kernel|core, default is app
+	SnapType string
+	// Modes in which the snap must be made available
+	Modes []string
+	// DefaultChannel is the initial tracking channel, default is stable
+	DefaultChannel string
+	// Track is a locked track for the snap, if set DefaultChannel cannot be
+	Track string
+	// Presence is one of: required|optional
+	Presence string
+}
+
+// SnapName implements naming.SnapRef.
+func (s *ModelSnap) SnapName() string {
+	return s.Name
+}
+
+// ID implements naming.SnapRef.
+func (s *ModelSnap) ID() string {
+	return s.SnapID
+}
+
+type modelSnaps struct {
+	Base   *ModelSnap
+	Gadget *ModelSnap
+	Kernel *ModelSnap
+	Snaps  []*ModelSnap
+}
+
+func (ms *modelSnaps) list() (allSnaps []*ModelSnap, allRequiredSnaps []naming.SnapRef, numBootSnaps int) {
+	addSnap := func(snap *ModelSnap, bootSnap int) {
+		if snap == nil {
+			return
+		}
+		numBootSnaps += bootSnap
+		allSnaps = append(allSnaps, snap)
+		if snap.Presence == "required" {
```

Comment:
Suggestion: If `Presence` was a distinct type we could have `snap.Presence.IsRequired()` that is only implemented in one spot, therefore avoiding chance of typos.

Reply:
we can think of that once the design is more stable

Reply:
added TODO